### PR TITLE
Use pathlib to manipulate paths and files

### PR DESCRIPTION
**Pull request by eumiro**
The project uses different ways to manipulate paths and files. This is an attempt to replace all `os.path.*`, `glob.*`, and `open` calls with `pathlib.Path` objects and methods that make the directory tree structure clearer and that allow better manipulation of disk-related operations.

### CODE REVIEW

**Comment by ultrabug** [comment]
Hey @eumiro, I hope you're doing good, I missed seeing you at EuroPython this year :/

Thanks a lot for this big contribution which makes a lot of sense, I shall review it properly and will get back to you :+1:

**Comment by eumiro** [comment]
Glad to hear that, @ultrabug .
Travis says there are problems in Python 3.5, but that is caused by the f-strings, which I completely ignored. I can replace them by `"".format` in my PR. Or are you considering dropping the Python 3.5 support in the near™ future?

**Comment by ultrabug** [comment]
@eumiro python3.5 reached EOL more than two months ago, I guess we could start getting rid of it

I think we should get a proper PR to update the setup.py to remove python3.5 and then we can proceed with yours?

We'll then need another PR with the run of pyupgrade to transform everything else while we're at it

Sounds like a plan?

**Comment by eumiro** [comment]
Great. See #1979

**Comment by ultrabug** [comment]
Cool :+1:

File: py3status/modules/file_status.py [hunk 1]
```diff
@@ -107,7 +104,7 @@ def post_config_hook(self):
         # convert str to list + expand path
         if not isinstance(self.paths, list):
             self.paths = [self.paths]
-        self.paths = list(map(expanduser, self.paths))
+        self.paths = [path.expanduser() for path in self.paths]
```
Comment (ultrabug) [inline py3status/modules/file_status.py:107]: I may have read through this too quick but how are the paths in self.path magically converted to Path so you can use expanduser() like this?

Comment (eumiro) [reply py3status/modules/file_status.py]: That's what reviews are for :+1: Thank you for checking it.

Fixed to

```
        self.paths = [Path(path).expanduser() for path in self.paths]
```

Is there actually no test coverage run in tox?

Comment (ultrabug) [reply py3status/modules/file_status.py]: I think this mistake is present on some other files, I'll get through them again then

> Is there actually no test coverage run in tox?

not those ones no; we've always been willing to but..

File: py3status/module.py [hunk 2]
```diff
@@ -588,7 +586,7 @@ def load_methods(self, module, user_modules):
             # user provided modules take precedence over py3status provided modules
             if self.module_name in user_modules:
                 include_path, f_name = user_modules[self.module_name]
```
Comment (ultrabug) [inline py3status/module.py:588]: this returns Path objects right?

File: py3status/modules/google_calendar.py [hunk 3]
```diff
@@ -195,8 +194,8 @@ def post_config_hook(self):
         self.events = None
         self.no_update = False
 
-        self.client_secret = os.path.expanduser(self.client_secret)
-        self.auth_token = os.path.expanduser(self.auth_token)
+        self.client_secret = self.client_secret.expanduser()
```
Comment (ultrabug) [inline py3status/modules/google_calendar.py:197]: here this module is missing Path

File: py3status/modules/imap.py [hunk 4]
```diff
@@ -123,8 +122,8 @@ def post_config_hook(self):
         self.idle_thread = Thread()
 
         if self.client_secret:
-            self.client_secret = os.path.expanduser(self.client_secret)
-        self.auth_token = os.path.expanduser(self.auth_token)
+            self.client_secret = self.client_secret.expanduser()
```
Comment (ultrabug) [inline py3status/modules/imap.py:125]: this modules needs Path

File: py3status/py3.py [hunk 5]
```diff
@@ -1103,7 +1103,7 @@ def play_sound(self, sound_file):
             if cmd:
                 if cmd == "ffplay":
                     cmd = "ffplay -autoexit -nodisp -loglevel 0"
-                sound_file = os.path.expanduser(sound_file)
+                sound_file = sound_file.expanduser()
```
Comment (ultrabug) [inline py3status/py3.py:1106]: missing Path?

**Comment by eumiro** [comment]
Thanks @ultrabug for the review, I have fixed and pushed all your suggestions. I think the strings should be converted to `Path` as soon as possible, even in the user modules, which should be then clearly defined in the interface.

File: fastentrypoints.py [hunk 6]
```diff
@@ -82,21 +82,21 @@ def get_args(cls, dist, header=None):  # noqa: D205,D400
 
 
 def main():
-    import os
     import re
     import shutil
     import sys
+    from pathlib import Path
```
Comment (lasers) [inline fastentrypoints.py:88]: Suggestion: Change this upstream. We can get it later along with any other changes.

**Review by ultrabug — CHANGES_REQUESTED** [review]
That's the last comment I have then we're good to go!

Sorry for the time it took me to get back to you mate

File: py3status/modules/file_status.py [hunk 7]
```diff
@@ -117,7 +117,7 @@ def post_config_hook(self):
 
     def file_status(self):
         # init data
-        paths = sorted([files for path in self.paths for files in glob(path)])
+        paths = sorted(files for path in self.paths for files in path.iterdir())
```
[the remark this answers is not in the document]
Comment (eumiro) [reply py3status/modules/file_status.py]: You're right, thank you for the review. I have now changed it to `path.parent.glob(path.name)` which should work for wildcards like `/a/b/*.d`. Do you think that would solve the problem?

[the remark this answers is not in the document]
Comment (ultrabug) [reply py3status/modules/file_status.py]: Yep I guess it should now, cheers

**Comment by ultrabug** [comment]
@eumiro merged and tested, it's broken... my review wasn't enough obviously !

**Comment by ultrabug** [comment]
Ok I believe I tackled most of them and master is now working as intended

**Comment by eumiro** [comment]
Thank you for the quick bugfixes and sorry for introducing those bugs. Can these cases be added to tests?

**Comment by ultrabug** [comment]
The whole test suite is to be reworked so I guess we should not waste our time in the v3 branch

I'll make sure some proper testing methodology is done on the version 4 I'm working on tho

**Comment by X-dark** [comment]
Seems there is still some issues:

```
 Traceback (most recent call last):
   File "/build/py3status-git/src/py3status/setup.py", line 14, in <module>
     sys.path.remove(module_path)
 ValueError: list.remove(x): x not in list
```

when trying to build with `python setup.py install --root="$pkgdir/" --optimize=1`

**Comment by ultrabug** [comment]
Thanks for reporting @X-dark

**Pull request merged**